### ISSUE TITLE
Ability to update from a distributed P2P network.

### ISSUE THREAD

**Issue by puneetse** [clearlinux/distribution#958]
It would be neat if CL could offer a choice to obtain update content from a P2P network hosted by willing CL clients. I'd love to get to the point where CL updates saturate my network pipes.  

The signing of content could ensure that the intended update content is being distributed.
Versioning/checks for updates could continue to be centralized to ensure clients aren't receiving stale information. 

Windows offers a similar feature, that can be enabled for LAN or WAN and has a controllable bandwidth throttle.

**Comment by lebensterben** [comment]
I agree.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by fenrus75** [comment]
P2P is .... kinda iffy.

our CDN somehow manages to suck. Plan to fix that is well in execution; but I think the right first step is to fix the CDN situation before revamping the whole software stack


note that we have a fallback CDN that might be worth seeing if it is faster

**Comment by fenrus75** [comment]
(P2P is iffy in that many ISPs do pretty nasty shaping things to it)

**Comment by ahkok** [comment]
Not to mention many P2P protocols rely on `UDP` which means it's wholly unusable in corporate networks.

**Comment by lebensterben** [comment]
> Not to mention many P2P protocols rely on `UDP` which means it's wholly unusable in corporate networks.

Restrictions on `UDP` should not be a big deal because there're ways to make `UDP` traffic looks like `TCP`.
Check this https://github.com/wangyu-/udp2raw-tunnel

In some parts of the world, P2P might be a better choice than CDN servers because the local authority may apply bandwidth throttling or even block the server. Surely you have an option to setup a local mirror site but then you'll have to remove anything they don't like from your server.

**Comment by puneetse** [comment]
> I think the right first step is to fix the CDN situation before revamping the whole software stack
Agreed, this isn't a solution to those CDN issues. I view a P2P distro update solution to be a capability that augments the update mechanism, with a core CDN as a reliable default.  

> (P2P is iffy in that many ISPs do pretty nasty shaping things to it)
Sure, but there are also networks where that is not a concern and those traffic shaping policies are not transparent. The choice to opt-in to a P2P solution should be the users.
